src/vs/editor/contrib/codeAction/browser/codeActionMenu.ts

```diff
@@ -50,34 +63,297 @@ export interface CodeActionShowOptions {
+		// data.detail.textContent = detail;
+
+		if (!isEnabled) {
+			data.root.classList.add('option-disabled');
+			data.root.style.backgroundColor = 'transparent !important';
+			data.root.style.color = 'rgb(204, 204, 204, 0.5)';
+			data.root.style.cursor = 'default';
+		} else {
+			data.root.classList.remove('option-disabled');
+		}
+
+		if (isSeparator) {
+			data.root.classList.add('separator');
+			data.root.style.height = '10px';
+		}
+
+	}
+	disposeTemplate(templateData: ICodeActionMenuTemplateData): void {
+		templateData.disposables = dispose(templateData.disposables);
+	}
+}
+
+export class CodeActionMenu extends Disposable implements IEditorContribution {
+	private readonly _ctxMenuWidgetIsFocused?: IContextKey<boolean>;
+	private readonly editor: ICodeEditor;
 	private readonly _showingActions = this._register(new MutableDisposable<CodeActionSet>());
+	private readonly _disposables = new DisposableStore();
+	private readonly _onDidHideContextMenu = new Emitter<void>();
+	private codeActionList!: List<ICodeActionMenuItem>;
+	private options: ICodeActionMenuItem[] = [];
+	private _visible: boolean = false;
+	readonly onDidHideContextMenu = this._onDidHideContextMenu.event;
+	private _ctxMenuWidgetVisible!: IContextKey<boolean>;
+	private viewItems: ICodeActionMenuItem[] = [];
+	private focusedEnabledItem!: number;
+	private currSelectedItem!: number;
+
+	public static readonly ID: string = 'editor.contrib.codeActionMenu';
+
+	public static get(editor: ICodeEditor): CodeActionMenu | null {
+		return editor.getContribution<CodeActionMenu>(CodeActionMenu.ID);
+	}
 
 	private readonly _keybindingResolver: CodeActionKeybindingResolver;
+	listRenderer: any;
 
 	constructor(
 		private readonly _editor: ICodeEditor,
 		private readonly _delegate: CodeActionWidgetDelegate,
 		@IContextMenuService private readonly _contextMenuService: IContextMenuService,
 		@IKeybindingService keybindingService: IKeybindingService,
 		@ILanguageFeaturesService private readonly _languageFeaturesService: ILanguageFeaturesService,
-		@ITelemetryService private readonly _telemetryService: ITelemetryService
+		@ITelemetryService private readonly _telemetryService: ITelemetryService,
+		@IThemeService _themeService: IThemeService,
+		@IConfigurationService private readonly _configurationService: IConfigurationService,
+		@IContextViewService private readonly _contextViewService: IContextViewService,
+		@IContextKeyService private readonly _contextKeyService: IContextKeyService,
 	) {
 		super();
 
+		this.editor = _editor;
 		this._keybindingResolver = new CodeActionKeybindingResolver({
 			getKeybindings: () => keybindingService.getKeybindings()
 		});
+
+		this._ctxMenuWidgetVisible = Context.Visible.bindTo(_contextKeyService);
 	}
 
 	get isVisible(): boolean {
 		return this._visible;
 	}
 
+	private isCodeActionWidgetEnabled(model: ITextModel): boolean {
+		return this._configurationService.getValue('editor.econtrib.codeAction.enabled', {
```

Comment:
Should be `editor.experimental.codeActionWidget.enabled`

Also can you extract the name to a constant to avoid problems like this?